vowpalwabbit/memory_tree.cc

```diff
@@ -0,0 +1,1358 @@
+                    ec.l.multi = mc;
+                else{
+                    ec.pred.multilabels = preds;
+                    ec.l.multilabels = multilabels;
+                }
+                ec.weight = ec_input_weight; //restore the original weight
+            }
+            else{ //if it's a leaf node:
+                float weight = 1.f; //float(path_to_leaf.size());
+                if (b.learn_at_leaf == true)
+                    learn_at_leaf_random(b, base, cn, ec, weight); //randomly sample one example, query reward, and update leaf learner
+
+                if (b.oas == true)
+                    train_one_against_some_at_leaf(b,base, cn, ec);
+            }
+        }
+        path_to_leaf.delete_v();
+    }
+
+    void insert_example_hal(memory_tree& b, single_learner& base, const uint32_t& ec_array_index, example& ec) 
+    {
+        //insert_example_without_ips(b, base, ec_array_index, true);
+        single_query_and_learn(b, base, ec_array_index, ec);
+        //multiple_query_learn_and_final_insert(b, base, ec_array_index, ec);
+    }
+
+    //node here the ec is already stored in the b.examples, the task here is to rout it to the leaf, 
+    //and insert the ec_array_index to the leaf.
+    void insert_example(memory_tree& b, single_learner& base, const uint32_t& ec_array_index, bool fake_insert = false)
+    {
+        uint32_t cn = 0; //start from the root.
+        while(b.nodes[cn].internal == 1) //if it's internal node:
+        {   
+            //predict and train the node at cn.
+            float router_pred = train_node(b, base, *b.examples[ec_array_index], cn); 
+            uint32_t newcn = descent(b.nodes[cn], router_pred); //updated nr or nl
+            cn = newcn; 
+        }
+    
+        if (b.oas == true)  //if useing oas as inference procedure, we just train oas here, as it's independent of the memory unit anyway'
+            train_one_against_some_at_leaf(b, base, cn, *b.examples[ec_array_index]);
+
+        if((b.nodes[cn].internal == -1) && (fake_insert == false)) //get to leaf:
+        {   
+            b.nodes[cn].examples_index.push_back(ec_array_index);
+            if (b.nodes[cn].examples_index.size() > b.max_ex_in_leaf)
+            {
+                b.max_ex_in_leaf = b.nodes[cn].examples_index.size();
+            }
+            float leaf_pred = train_node(b, base, *b.examples[ec_array_index], cn); //tain the leaf as well.
+            descent(b.nodes[cn], leaf_pred); //this is a faked descent, the purpose is only to update nl and nr of cn
+
+            //if the number of examples exceeds the max_leaf_examples, and not reach the max_nodes - 2 yet, we split:
+            if((b.nodes[cn].examples_index.size() >= b.max_leaf_examples) && (b.nodes.size() + 2 <= b.max_nodes)){
+                split_leaf(b, base, cn); 
+            }
+        }
+    }
+
+    void experience_replay(memory_tree& b, single_learner& base)
+    {
+      //return; // TODO turn this back on
```

Comment:
What's the TODO status here?

Reply:
deleted this comment, it was for debug purpose.